flang/lib/Lower/IO.cpp

```diff
@@ -523,6 +523,22 @@ static mlir::FuncOp getInputFunc(mlir::Location loc, fir::FirOpBuilder &builder,
   return getIORuntimeFunc<mkIOKey(InputDescriptor)>(loc, builder);
 }
 
+/// Interpret addr as a C++ bool (1 byte), and then set its value to the
+/// equivalent Fortran logical. This needs to be called for logical that were
+/// passed as bool& to the runtime and whose upper bytes, if any, may not be
+/// correct anymore.
```

Comment:
This reads like a hack to workaround bugs in the runtime. Why is the runtime writing garbage to LOGICAL memory locations?

Reply:
The runtime is not writing garbage, it is just expecting a `bool`, and not a `logical` here to avoid duplicating the entry points. We previously wrongly pass a `fir.ref<logical>` as being an `fir.ref<i1>` here, this worked OK in cases where only the lowest bit matters.

What we are doing here is basically just to use the `logical` storage as a `bool` storage and to interpret it back from `i1` to `logical` afterwards. We could also have made an `i1` temp alloca, pass that to the runtime, and interpret it back to the logical. It was easier to use the logical storage and it saved an alloca.

Reply:
But the lsbit is really the only thing that matters given C++ will write a 0 or a 1 to the lsbyte.

I think it might be clearer in the comment to explicitly state that this is load and store effectively (sign or zero) extends the lsbyte into the full LOGICAL value storage, as the runtime is unaware of the LOGICAL value's actual bit width.

Reply:
Thanks, I updated the comment as suggested.